Test playbook using molecule

sleiner opened the pull request:
# Proposed Changes

- The test suite uses molecule instead of interacting with Vagrant and a custom Python script directly.

## Checklist

- [x] Tested locally
- [x] Ran `site.yml` playbook
- [x] Ran `reset.yml` playbook
- [x] Did not add any unnecessary changes
- [x] 🚀

sleiner commented on molecule/README.md:
````diff
@@ -0,0 +1,67 @@
+- [Vagrant](https://www.vagrantup.com/downloads)
+- [VirtualBox](https://www.virtualbox.org/wiki/Downloads)
+
+### Set up VirtualBox networking on Linux and macOS
+
+_You can safely skip this if you are working on Windows._
+
+Furthermore, the test cluster uses the `192.168.30.0/24` subnet which is [not set up by VirtualBox automatically](https://www.virtualbox.org/manual/ch06.html#network_hostonly).
+To set the subnet up for use with VirtualBox, please make sure that `/etc/vbox/networks.conf` exists and that it contains this line:
+
+```
+* 192.168.30.0/24`
+```
+
````
@timothystewart6 we could get rid of this step by changing over the sample inventory's IP addresses from `192.168.30.0/24` to something in `192.168.56.0/21`, e.g. `192.168.60.0/24`. What do you think?

sleiner replied:
Alright, sounds sensible 👍🏻 Previously, I was not sure if this was chosen completely arbitrarily 😄

sleiner commented:
@timothystewart6 should the PR checklist be changed?

twistedgrim commented:
Excellent.... nice job!

timothystewart6 commented:
> 

I think so now that we have CI running.  I think we can do it in a separate PR!

timothystewart6 commented:
Why don't we see the output of molecule when testing in the `Test with molecule` step?  Am I missing something?

timothystewart6 commented:
Note to self, I will have to update this post. https://github.com/techno-tim/techno-tim.github.io/blob/master/_posts/2022-03-26-k3s-etcd-ansible.md

sleiner commented:
@timothystewart6 

> I think so now that we have CI running. I think we can do it in a separate PR!

I'm missing a bit of context here 😅 _What_ could be done in a separate PR?

---

> Why don't we see the output of molecule when testing in the Test with molecule step? Am I missing something?

To debug some file permission issue on the runner itself more quickly, I had replaced the molecule invocation. The issue is now fixed and molecule is back in place 😊

timothystewart6 commented:
> Oh, sorry, I thought I quoted this!  I mean that the PR template can be done in another PR!

sleiner commented:
@timothystewart6
have you seen these types of errors before?
```
TASK [k3s/post : Wait for metallb controller to be running] ********************
[496](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:504)
  failed: [control1] (item=control1) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.729871", "end": "2022-09-02 15:06:12.479623", "item": "control1", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:10.749752", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
[497](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:505)
  failed: [control1] (item=control2) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.886031", "end": "2022-09-02 15:06:17.390877", "item": "control2", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:15.504846", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
[498](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:506)
  failed: [control1] (item=control3) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:02.417555", "end": "2022-09-02 15:06:23.943977", "item": "control3", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:21.526422", "stderr": "error: the server doesn't have a resource type \"deployment\"", "stderr_lines": ["error: the server doesn't have a resource type \"deployment\""], "stdout": "", "stdout_lines": []}
```

timothystewart6 commented:
> @timothystewart6 have you seen these types of errors before?
> 
> ```
> TASK [k3s/post : Wait for metallb controller to be running] ********************
> [496](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:504)
>   failed: [control1] (item=control1) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.729871", "end": "2022-09-02 15:06:12.479623", "item": "control1", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:10.749752", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
> [497](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:505)
>   failed: [control1] (item=control2) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.886031", "end": "2022-09-02 15:06:17.390877", "item": "control2", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:15.504846", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
> [498](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:506)
>   failed: [control1] (item=control3) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:02.417555", "end": "2022-09-02 15:06:23.943977", "item": "control3", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:21.526422", "stderr": "error: the server doesn't have a resource type \"deployment\"", "stderr_lines": ["error: the server doesn't have a resource type \"deployment\""], "stdout": "", "stdout_lines": []}
> ```
> 
> I could build a retry in, but the API server not being available (and afterwards not knowing `deployment`s) seems super weird - especially since at this point we have already used the API server quite a bit 🤔

I just pushed up a fix, not sure if this will actually fix your issue because it still works locally and in my test cluster without this.  Might be worth merging in.

timothystewart6 commented:
(I updated the branch)

timothystewart6 commented on inventory/sample/hosts.ini:
```diff
@@ -1,11 +1,11 @@
-192.168.30.38
-192.168.30.39
-192.168.30.40
+control1 ansible_host=192.168.30.38
```
I am wondering if it just makes sense to create another collection rather than update this collection with variables.  Is it possible to create a `test.ini`  and use that for testing only without adding these variables to `host.ini`?  Asking because 99.99% of users won't ever use the tests however now they have to discern test variables.

timothystewart6 replied:
It's not a blocker, just asking the question :)

twistedgrim replied:
I would agree. A separate test.ini maybe in a tests directory would be good. Maybe even just put it in the .github directory.

sleiner replied:
Good point, I have dropped that part of the patch 👍🏻

timothystewart6 replied:
Thank you!

timothystewart6 commented:
> > @timothystewart6 have you seen these types of errors before?
> > ```
> > TASK [k3s/post : Wait for metallb controller to be running] ********************
> > [496](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:504)
> >   failed: [control1] (item=control1) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.729871", "end": "2022-09-02 15:06:12.479623", "item": "control1", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:10.749752", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
> > [497](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:505)
> >   failed: [control1] (item=control2) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.886031", "end": "2022-09-02 15:06:17.390877", "item": "control2", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:15.504846", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
> > [498](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:506)
> >   failed: [control1] (item=control3) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:02.417555", "end": "2022-09-02 15:06:23.943977", "item": "control3", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:21.526422", "stderr": "error: the server doesn't have a resource type \"deployment\"", "stderr_lines": ["error: the server doesn't have a resource type \"deployment\""], "stdout": "", "stdout_lines": []}
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I could build a retry in, but the API server not being available (and afterwards not knowing `deployment`s) seems super weird - especially since at this point we have already used the API server quite a bit 🤔
> 
> I just pushed up a fix, not sure if this will actually fix your issue because it still works locally and in my test cluster without this. Might be worth merging in.

This might have fixed it.

sleiner commented:
> > > @timothystewart6 have you seen these types of errors before?
> > > ```
> > > TASK [k3s/post : Wait for metallb controller to be running] ********************
> > > [496](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:504)
> > >   failed: [control1] (item=control1) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.729871", "end": "2022-09-02 15:06:12.479623", "item": "control1", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:10.749752", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
> > > [497](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:505)
> > >   failed: [control1] (item=control2) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:01.886031", "end": "2022-09-02 15:06:17.390877", "item": "control2", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:15.504846", "stderr": "Error from server (ServiceUnavailable): apiserver not ready", "stderr_lines": ["Error from server (ServiceUnavailable): apiserver not ready"], "stdout": "", "stdout_lines": []}
> > > [498](https://github.com/techno-tim/k3s-ansible/runs/8158547698?check_suite_focus=true#step:7:506)
> > >   failed: [control1] (item=control3) => {"ansible_loop_var": "item", "changed": false, "cmd": ["kubectl", "wait", "deployment", "-n", "metallb-system", "controller", "--for", "condition=Available=True", "--timeout=60s"], "delta": "0:00:02.417555", "end": "2022-09-02 15:06:23.943977", "item": "control3", "msg": "non-zero return code", "rc": 1, "start": "2022-09-02 15:06:21.526422", "stderr": "error: the server doesn't have a resource type \"deployment\"", "stderr_lines": ["error: the server doesn't have a resource type \"deployment\""], "stdout": "", "stdout_lines": []}
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > I could build a retry in, but the API server not being available (and afterwards not knowing `deployment`s) seems super weird - especially since at this point we have already used the API server quite a bit 🤔
> > 
> > 
> > I just pushed up a fix, not sure if this will actually fix your issue because it still works locally and in my test cluster without this. Might be worth merging in.
> 
> This might have fixed it, but now there's an error further down in the pipeline :)

That makes sense, I forgot to commit something 😄 Retrying it now...

sleiner commented:
fyi: I just realized that the overrides are specific to the `default` scenario and modified the respective commit :-)

twistedgrim commented on molecule/default/overrides.yml:
```diff
@@ -0,0 +1,11 @@
+        flannel_iface: eth1
+
+        # The test VMs might be a bit slow, so we give them more time to join the cluster:
+        retry_count: 45
```
Instead of a ton of retries do you want to do a wait before?

sleiner replied:
IMO, that is out of scope for this PR.

twistedgrim reviewed (commented):
Looks nice. Great work. I like the host to host name changes.

timothystewart6 reviewed (commented):
This looks good!  Thank you so much!

timothystewart6 commented:
Are you ok if I merge @sleiner ?

sleiner commented:
Yes 🙌🏻
